src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -761,7 +777,7 @@ private static unsafe string JoinCore<T>(char* separator, int separatorLength, I
             }
         }
 
-        private static unsafe string JoinCore(char* separator, int separatorLength, string?[] value, int startIndex, int count)
+        private static unsafe string JoinCore(char* separator, int separatorLength, IList<string?> value, int startIndex, int count)
```

Comment:
Have you benchmarked this?  In theory this could regress existing usage with arrays, as now every access is via an interface call.

Reply:
I pinged John offline with tips on using our performance repo's benchmarking suite. John - if you have any other questions re: benchmarking feel free to drop them in the conversation here.

Reply:
It does appear to be slower, do you see a way to maintain the performance without duplicating this method?
|      Method | Toolchain |      Mean |    Error |   StdDev |    Median |       Min |       Max | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------ |---------- |----------:|---------:|---------:|----------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
|   ArrayTest |       new |  85.41 ns | 5.792 ns | 6.438 ns |  85.05 ns |  74.32 ns |  96.83 ns |  1.00 |    0.00 | 0.0185 |     - |     - |      80 B |
|   ArrayTest |       old |  49.66 ns | 1.659 ns | 1.775 ns |  49.87 ns |  47.21 ns |  52.95 ns |  0.58 |    0.04 | 0.0185 |     - |     - |      80 B |

Reply:
If we wanted to special-case `string[]` and `List<string>` we could convert both to `ReadOnlySpan<string>`, which should have array-equivalent performance characteristics. That doesn't help other concrete types that implement `IList<string>`, though. (Is this common?)

Reply:
I like Levi's suggestion (and probably just for `List<string>`, since I think it's unlikely you're in a situation where you're getting a `string[]` strongly typed as an `IEnumerable<string>` and care about this level of optimization).  Is there another input that's inspiring you here?

Reply:
FWIW, the `ROS<string?>` solution would look something like this:

```cs
private static void JoinCore(..., ReadOnlySpan<string?> values, ...) { /* ... */ }

IEnumerable<string?> enumerable = ...;
if (enumerable is string?[] arr) { return JoinCore(..., arr, 0, arr.Length); }
else if (enumerable is List<string?> list) { return JoinCore(..., CollectionsMarshal.AsSpan(list), 0, list.Count); }
else { /* fall back to existing enum-based logic */ }
```

Reply:
> if (enumerable is string?[] arr) { return JoinCore(..., arr, 0, arr.Length); }

Is this check even necessary?  As noted, I think it's unlikely you're in a situation where you're getting a `string[]` strongly typed as an `IEnumerable<string>` and care about this level of optimization.  If you have a strongly-typed `string[]`, you'll bind to the existing `string[]`-based overload.

Reply:
At least in my teams case we often have both arrays and lists passed as IList. I'm now realizing that this isn't as performant as having it kept as T[] all the way back. I agree that having a string[] typed as a IEnumerable<string> is cases it's types as anything other than string[] there will be a significant perf improvement. I'm happy to remove the `enumerable is string?[] arr` case, let me know what you'd like to do.
|       Method |        Job |  Toolchain |      Mean |     Error |    StdDev |    Median |       Min |       Max | Ratio | RatioSD |
|------------- |----------- |----------- |----------:|----------:|----------:|----------:|----------:|----------:|------:|--------:|
|    ArrayTest | Job-EOKGEP |        old |  49.50 ns |  2.105 ns |  2.252 ns |  49.43 ns |  45.51 ns |  53.95 ns |  1.00 |    0.00 |
|    ArrayTest | Job-TTJIGM |        new |  49.90 ns |  1.592 ns |  1.833 ns |  50.21 ns |  46.79 ns |  52.93 ns |  1.01 |    0.06 |
|              |            |            |           |           |           |           |           |           |       |         |
|         List | Job-EOKGEP |        old | 125.01 ns |  4.675 ns |  5.196 ns | 123.35 ns | 118.42 ns | 137.19 ns |  1.00 |    0.00 |
|         List | Job-TTJIGM |        new |  65.14 ns |  2.381 ns |  2.742 ns |  65.01 ns |  60.65 ns |  70.58 ns |  0.52 |    0.02 |
|              |            |            |           |           |           |           |           |           |       |         |
| ArrayAsIList | Job-EOKGEP |        old |  88.90 ns |  2.787 ns |  2.983 ns |  88.37 ns |  84.83 ns |  95.60 ns |  1.00 |    0.00 |
| ArrayAsIList | Job-TTJIGM |        new |  54.17 ns |  1.419 ns |  1.577 ns |  54.04 ns |  51.30 ns |  57.32 ns |  0.61 |    0.03 |
|              |            |            |           |           |           |           |           |           |       |         |
|  IEnumerable | Job-EOKGEP |        old | 206.12 ns | 17.513 ns | 20.168 ns | 202.55 ns | 177.56 ns | 240.59 ns |  1.00 |    0.00 |
|  IEnumerable | Job-TTJIGM |        new | 190.81 ns |  6.470 ns |  7.451 ns | 192.75 ns | 177.33 ns | 201.28 ns |  0.93 |    0.09 |

Reply:
These numbers look good to me. I don't really have any other feedback on the code. @stephentoub - did the latest iteration address your concerns?

(There's a little more cleanup we can do, such as nixing the _startIndex_ and _count_ parameters and removing some uses of _unsafe_, but that's best left for a followup PR.)